### ISSUE TITLE
eslint path problems

### ISSUE THREAD

**Issue by acthp** [scrooloose/syntastic#1692]
As noted in another issue, global eslint will not run the eslint version installed in a project. The recommended way to run the correct version is via npm, like 'npm run lint', after creating a lint script directive in package.json, like {'lint': 'eslint'}.

However, setting javascript_eslint_exec to 'npm run lint' in syntastic results in syntastic not finding the executable. Setting it to 'npm' and javascript_eslint_args to 'run lint --' results in command line

    npm -f compat run lint -- foo.js

which is wrong. It needs to be

    npm run lint -- -f compat foo.js

The 'args_before' call botches the order.

We need better control of the command line to make it work.

**Comment by lcd047** [comment]
`javascript_eslint_exe`

**Comment by acthp** [comment]
Thank you!

    let g:syntastic_javascript_eslint_exe = 'npm run eslint --'

This works.

**Comment by gurpreetatwal** [comment]
If you don't want to add a custom task to your package scripts the following setting will always run the locally installed eslint

```
let g:syntastic_javascript_eslint_exe = '$(npm bin)/eslint
```

**Comment by stuartsan** [comment]
Note that may help someone else: for me the approaches described by @acthp and @gurpreetatwal, which are aimed at using a project's local version of `eslint`, did not work until I _also_ had `eslint` installed globally (`npm install -g eslint`).

**Comment by lcd047** [comment]
@stuartsan Just point `*_exec` (not `*_exe`) to a valid binary:
```vim
let g:syntastic_javascript_eslint_exec = '/bin/ls'
```

**Comment by stuartsan** [comment]
Ahh gtk, thanks @lcd047 !

**Comment by kogakure** [comment]
I wrote a solution to use the local node_module script or fallback to the global installation, when not available:

```vim
let g:syntastic_javascript_eslint_exe = '[ -f $(npm bin)/eslint ] && $(npm bin)/eslint || eslint'
```

**Comment by lcd047** [comment]
@kogakure Alternatively, write a wrapper script and point `g:syntastic_javascript_eslint_exec` to it.

**Comment by ezmiller** [comment]
I have tried a couple of these examples for getting syntastic to use the local eslint but to no avail. The above: 

`let g:syntastic_javascript_eslint_exe = '[ -f $(npm bin)/eslint ] && $(npm bin)/eslint || eslint'

Led to output like this:

```
syntastic: 7.221055: g:syntastic_version = '3.8.0-14 (Vim 704, Neovim, Darwin)'
syntastic: 7.221276: &shell = '/bin/zsh', &shellcmdflag = '-c', &shellpipe = '2>&1
| tee', &shellquote = '', &shellredir = '>%s 2>&1', &shelltemp = 1, &shellxquote =
 '', &autochdir = 0, &shellxescape = ''
syntastic: 7.222215: UpdateErrors (auto): default checkers
syntastic: 7.222640: CacheErrors: default checkers
syntastic: 7.223323: g:syntastic_aggregate_errors = 0
syntastic: 7.223588: getcwd() = '/Users/ethan/Projects/motologic/enzo'
syntastic: 7.234459: CacheErrors: Checker javascript/eslint is not available
syntastic: 7.234702: CacheErrors: no checkers available for javascript.jsx
```

Like @stuartsan, I've only been able to get eslint to work with syntastic if I have it installed globally, which is not ideal.

**Comment by lcd047** [comment]
> Like @stuartsan, I've only been able to get eslint to work with syntastic if I have it installed globally, which is not ideal.

Like @stuartsan, you can just point `*_exec` (not `*_exe`) to a valid binary:
``` vim
let g:syntastic_javascript_eslint= '/bin/ls'
```
That's because `*_exec ` has to point to an executable file before syntastic would consider enabling the checker.  When you install `eslint` locally there is no `eslint` in your `$PATH`, and thus you get:
````
syntastic: 7.234459: CacheErrors: Checker javascript/eslint is not available
```

**Comment by lcd047** [comment]
> Like @stuartsan, I've only been able to get eslint to work with syntastic if I have it installed globally, which is not ideal.

@ezmiller Like @stuartsan, you can just point `*_exec` (not `*_exe`) to a valid binary:
``` vim
let g:syntastic_javascript_eslint= '/bin/ls'
```
That's because `*_exec ` has to point to an executable file before syntastic would consider enabling the checker.  When you install `eslint` locally there is no `eslint` in your `$PATH`, and thus you get:
````
syntastic: 7.234459: CacheErrors: Checker javascript/eslint is not available
```

**Comment by ezmiller** [comment]
@lcd047 I tried the `*_exec` as well with this output:

```
syntastic: 7.757566: g:syntastic_version = '3.8.0-14 (Vim 704, Neovim, Darwin)'
syntastic: 7.757836: &shell = '/bin/zsh', &shellcmdflag = '-c', &shellpipe = '2>&1
| tee', &shellquote = '', &shellredir = '>%s 2>&1', &shelltemp = 1, &shellxquote =
 '', &autochdir = 0, &shellxescape = ''
syntastic: 7.758561: UpdateErrors (auto): default checkers
syntastic: 7.758999: CacheErrors: default checkers
syntastic: 7.759670: g:syntastic_aggregate_errors = 0
syntastic: 7.759912: getcwd() = '/Users/ethan/Projects/motologic/enzo'
syntastic: 7.779182: CacheErrors: Checker javascript/eslint is not available
syntastic: 7.779530: CacheErrors: no checkers available for javascript.jsx
syntastic: 7.779716: aggregated: {'_sorted': 0, '_name': '', '_owner': 3, '_column
s': 1, '_rawLoclist': []}
```

I also wanted to see if for some reason `npm bin` wasn't working when vim ran (actually I'm running neovim), so I tried:

`:echo syntastic#util#system('$(npm bin)/eslint --version')`

And it worked. So I'm not sure what might be going wrong here. Thoughts?

**Comment by lcd047** [comment]
You also have to set `g:syntastic_javascript_eslint_generic` to 1.  It isn't like `node` is a lesser mess, you know.

**Comment by ezmiller** [comment]
@lcd047 thanks. Is this stuff in the docs somewhere? You are right node is also a mess.

**Comment by ezmiller** [comment]
@lcd047 adding that line doesn't get things working unfortunately. The error output seems virtually identical:

```
 "~/Projects/motologic/enzo/client/app/bundles/Moto/components/ListSearchResults/ListSearch
Results.jsx" 86L, 2366C
syntastic: 8.056502: g:syntastic_version = '3.8.0-14 (Vim 704, Neovim, Darwin)'
syntastic: 8.056757: &shell = '/bin/zsh', &shellcmdflag = '-c', &shellpipe = '2>&1| tee',
&shellquote = '', &shellredir = '>%s 2>&1', &shelltemp = 1, &shellxquote = '', &autochdir
= 0, &shellxescape = ''
syntastic: 8.057520: UpdateErrors (auto): default checkers
syntastic: 8.057959: CacheErrors: default checkers
syntastic: 8.058722: g:syntastic_aggregate_errors = 0
syntastic: 8.058954: getcwd() = '/Users/ethan/Projects/motologic/enzo'
syntastic: 8.076664: CacheErrors: Checker javascript/eslint is not available
syntastic: 8.076996: CacheErrors: no checkers available for javascript.jsx
syntastic: 8.077217: aggregated: {'_sorted': 0, '_name': '', '_owner': 1, '_columns': 1, '
_rawLoclist': []}
"~/.vimrc" 203L, 6066C
syntastic: 29.017198: g:syntastic_version = '3.8.0-14 (Vim 704, Neovim, Darwin)'
syntastic: 29.017469: &shell = '/bin/zsh', &shellcmdflag = '-c', &shellpipe = '2>&1| tee',
 &shellquote = '', &shellredir = '>%s 2>&1', &shelltemp = 1, &shellxquote = '', &autochdir
 = 0, &shellxescape = ''
syntastic: 29.018205: UpdateErrors (auto): default checkers
syntastic: 29.018742: CacheErrors: default checkers
syntastic: 29.019432: g:syntastic_aggregate_errors = 0
syntastic: 29.019695: getcwd() = '/Users/ethan/Projects/motologic/enzo'
syntastic: 29.024159: CacheErrors: Checker vim/vimlint is not available
syntastic: 29.024502: CacheErrors: no checkers available for vim
syntastic: 29.024668: aggregated: {'_sorted': 0, '_name': '', '_owner': 3, '_columns': 1,
'_rawLoclist': []}
```

**Comment by lcd047** [comment]
@ezmiller Setting `g:syntastic_debug` to 33 might tell you why.

**Comment by gareth** [comment]
From my testing, this approach (setting `generic` `eslint`) doesn't pass the `-f compact` option and I guess the errors aren't extracted properly

**Comment by gurpreetatwal** [comment]
hmm that's odd, what properties are you setting exactly? This solution was worked fine for me for as long as I used syntastic. (Using ALE these days)

**Comment by gareth** [comment]
That basically comes from [here](https://github.com/vim-syntastic/syntastic/blob/master/syntax_checkers/javascript/eslint.vim#L45) as far as I can see.

```
let g:syntastic_javascript_checkers  = ['eslint']
let g:syntastic_javascript_eslint_exe='$(npm bin)/eslint'
let g:syntastic_javascript_eslint_exec='/bin/ls'
let g:syntastic_javascript_eslint_generic = 1
```
leads to
```
syntastic: 0.206305: g:syntastic_version = '3.8.0-63 (Vim 800, Darwin, GUI)'
syntastic: 0.206450: &shell = '/bin/zsh', &shellcmdflag = '-c', &shellpipe = '2>&1| tee', &shellquote = '', &shellredir = '>%s 2>&1', &shelltemp = 1, &shellxquote = '', &autochdir = 0, &shellxescape = ''
syntastic: 0.206930: UpdateErrors (auto): default checkers
syntastic: 0.207296: CacheErrors: default checkers
syntastic: 0.208292: g:syntastic_aggregate_errors = 0
syntastic: 0.208441: $TERM = 'xterm-256color'
syntastic: 0.208528: $TMPDIR = '/var/folders/4v/jj7x0j1j2gnd8c9ztht3rr2r0000gn/T/'
syntastic: 0.208609: $PATH = '/usr/local/Caskroom/google-cloud-sdk/latest/google-cloud-sdk/bin:/usr/local/heroku/bin:/usr/local/share/npm/bin:/usr/local/bin:/usr/local/sbin:/Users/gareth/bin:/usr/local/var/rbenv/shims:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/MacGPG2/bin:/Users/gareth/.go/bin'
syntastic: 0.208711: getcwd() = '/Users/gareth/project'
syntastic: 0.214614: javascript/eslint: generic eslint, exec = '/bin/ls'
syntastic: 0.214850: CacheErrors: Invoking checker: javascript/eslint
syntastic: 0.215573: SyntasticMake: called with options: {'postprocess': ['guards'], 'errorformat': '%E%f: line %l\, col %c\, Error - %m,%W%f: line %l\, col %c\, Warning - %m', 'makeprg': '$(npm bin)/eslint app/assets/javascripts/components/direct_upload.js'}
syntastic: 1.794736: system: command run in 1.578781s
syntastic: 1.796100: getLocList: checker javascript/eslint returned 1
syntastic: 1.796379: getLocList: checker javascript/eslint run in 1.581400s
```

Note: `'makeprg': '$(npm bin)/eslint app/assets/javascripts/components/direct_upload.js'` doesn't have an `-f compact`. Maybe the "generic" approach doesn't want to assume that compact is available?

**Comment by gurpreetatwal** [comment]
Interestingly enough I never had the `generic` option set, this is what used to be in my vimrc for syntastic:

```
let g:syntastic_check_on_wq=1
let g:syntastic_auto_loc_list=2
let g:syntastic_javascript_checkers=['eslint']	
let g:syntastic_javascript_eslint_exe='eslint_d'
let g:syntastic_javascript_eslint_args=['--cache']
```

**Comment by lcd047** [comment]
> From my testing, this approach (setting `generic` `eslint`) doesn't pass the `-f compact` option and I guess the errors aren't extracted properly

The recommended approach is the one in the __first__ post above.  It takes a time investment of ~2 minutes to understand and apply correctly.  Most of the reminder of the thread is just people trying random other things they picked up elsewhere, hoping to avoid spending those ~2 minutes. _shrug_

**Comment by gareth** [comment]
@lcd047 The approach in the first post seems to rely on `eslint` being installed globally, so that the [`executable`](https://github.com/vim-syntastic/syntastic/blob/master/syntax_checkers/javascript/eslint.vim#L33) check - and the subsequent [`--version`](https://github.com/vim-syntastic/syntastic/blob/master/plugin/syntastic/checker.vim#L147) check can succeed. If the value of `_exec` isn't a route to a copy of `eslint` that you can pass `--version` to, then you have to use the `_generic` option - which as I pointed out, doesn't return errors in a format that Syntastic can use unless the `-f compact` option is passed.

Am I right that having a globally accessible install of `eslint` is required to use Syntastic's integration - even if that version is completely ignored by overriding it using `_exe`?

Maybe if you've used npm a lot, and maybe if you know vimscript, and maybe if you *wrote* Syntastic then it only takes 2 minutes to understand and apply these concepts, but for someone who *hasn't* made a thousand commits to this project I did spend a couple of hours actually trying to work out the execution flow before posting here. I just didn't detail everything I tried in my message because, well, who wants to read an essay? It's a little offputting to try asking the author of a plugin for help and basically be told "You're not clever enough to use my plugin" ☹️

**Comment by lcd047** [comment]
@gareth Ah, so the problem is `generic` is used to disable both the initial version check and the subsequent output format option.  Well that used to make sense, since the point of `generic` was to allow people to run the myriad `eslint`-derived linters within syntastic's `eslint`-checker, without having to add a corresponding myriad of syntastic checkers.

> Am I right that having a globally accessible install of `eslint` is required to use Syntastic's integration - even if that version is completely ignored by overriding it using `_exe`?

It isn't strictly _required_, you can also write a wrapper script that handles all those executable-finding contortions, and point syntastic's `exec` to it.  But yes, installing a global `eslint` would also work.

The other solution if you don't want to install a global `eslint` is the mess above, with `/bin/ls` and `generic`, and then add back `-f compact` via `g:syntastic_javascript_eslint_args`.  I'm not sure which one is uglier.

> It's a little offputting to try asking the author of a plugin for help and basically be told "You're not clever enough to use my plugin"

Sorry if you read it that way.  I suppose it was the result of my idle meditation on how everybody complains  about syntastic not being able to handle all this crap automatically, or at least smoothly, while nobody ever questions why `node` makes said crap necessary to begin with.  Oh well, as somebody else points out, there are better alternatives to syntastic these days.

**Comment by gareth** [comment]
Thanks for the considered reply, that's a great summary of my options.

Believe me, I also have a lot of opinions about the amount of crap that node forces on you 😉 But I think asking questions about large projects is even *harder* because of the weight of development behind those, and speaking up is pretty intimidating.

But that's not a topic for this thread, which I have a solution for now 😄

**Comment by timwis** [comment]
Hey folks, just wanted to let you know that since npm v5 ships with [npx](https://www.npmjs.com/package/npx) you can use this in place of `$(npm bin)`, ie:

```vim
let g:syntastic_javascript_eslint_exe = 'npx standard'
```

**Comment by ericandrewlewis** [comment]
To summarize: you have two options to get vim-syntastic to use a locally installed version of ESLint.

## With ESLint installed globally

Install ESLint globally. Syntastic won't use this version, it just checks it for a few things.

```
npm install -g eslint
```

In your `~.vimrc`:
```
" Enable the eslint checker
let g:syntastic_javascript_checkers=['eslint']
" Point the checker to the locally installed version of eslint
let g:syntastic_javascript_eslint_exe='$(npm bin)/eslint'
```

## Without global ESLint installed

In your `~.vimrc`:
```
" Enable the eslint checker
let g:syntastic_javascript_checkers=['eslint']
let g:syntastic_javascript_eslint_generic = 1
let g:syntastic_javascript_eslint_exec = '/bin/ls'
let g:syntastic_javascript_eslint_exe='$(npm bin)/eslint'
let g:syntastic_javascript_eslint_args='-f compact'
```

## Set g:syntastic_javascript_eslint_exec to a wrapper script

I'm not sure how this option would work but it's described above.

**Comment by lcd047** [comment]
@ericandrewlewis Please don't make authoritative statements until you understand exactly what's going on.  Setting `exe` to `$(npm bin)/eslint` breaks when you're switching projects from withing Vim.

**Comment by ericandrewlewis** [comment]
@lcd047 good point. If I open a file in a separate project ESLint is not run properly against it. If I change the working directory for that buffer and run `:SyntasticCheck`, that project's ESLint is invoked and the results show as expected. This solution depends on [automatically changing the working directory for each window](http://vim.wikia.com/wiki/Set_working_directory_to_the_current_file), which sounds like it can cause trouble with other plugins.

Alternatively I gave a shot to making a wrapper script which does the trick. To set this up:

In `~/bin/eslint-wrapper-for-vim-syntastic`
```bash
#!/bin/bash
if [ "$1" = "--version" ] 
	then
		echo '1.0.0'
		exit
fi

$(npm bin)/eslint $@
```

then in `.vimrc`

```
let g:syntastic_javascript_eslint_exec = '~/bin/eslint-wrapper-for-vim-syntastic'
```

**Comment by lcd047** [comment]
@ericandrewlewis Most checkers would work with `autochdir`, but not all.

A simple-minded wrapper as above still won't work, unless you change Vim's idea of current directory with `:lcd`.  A better solution would be to examine arguments passed to the script, identify the files among them (something like `if [ "$1" ] && [ -f "$1" ]; then ...`), and run `npm bin` from the `basename` of said files.  There's a reason why the thread above is so long. :smile:

Also, you don't need `/bin/bash`, plain `/bin/sh` is enough.  And you need to quote `$@`.

**Comment by kevinoid** [comment]
After trying several of the approaches in this issue, I settled on an [implementation](https://github.com/kevinoid/syntastic-use-node-bin#implementation) which uses executables in `node_modules/.bin` when present, supports multiple buffers in separate projects, and is reasonably generic and flexible.  I packaged it as a plugin ([kevinoid/syntastic-use-node-bin](https://github.com/kevinoid/syntastic-use-node-bin)) so others could use it easily.  I'd appreciate constructive feedback.  If there is any interest in including this feature in Syntastic, I would be happy to help merge it.

**Comment by lcd047** [comment]
@kevinoid I has been a while since I wrote the registry code, and I'm not entirely convinced what you're doing makes sense in all possible situations (among other things, the registry is initialized only when some checkers are actually requested).  But it's good to know it solves your problem. :smile:

**Comment by kevinoid** [comment]
@lcd047 Thanks for taking a look at it.  If you would like me to convince you (possibly by fixing cases I've overlooked in the process), I'm open to that.  Probably only worth our time if there is a chance it could be merged.  :smile:

**Comment by lcd047** [comment]
@kevinoid If your change requires the registry to be fully loaded for certain filetypes this slows down the first run of a checker, sometimes by a few seconds.  So no, I'm afraid I won't merge it. :smile:

Then again, I'm pretty sure I don't understand the problem you're trying to solve.  In particular, I see no reason to do all the work _after_ syntastic and the registry have been initialized, instead of doing it with a script (`make(1)` comes to mind) only once, when you start a new project.  Or perhaps whenever you add new files to the project.  You don't even have to do it at every run of Vim.  Trying to make it configure itself is what complicates your life exponentially, and it's about as opposed to syntastic's design as it gets.

**Comment by kevinoid** [comment]
@lcd047 I'm trying to have Syntastic use linter executables from `node_modules/.bin`, which is required for ESLint (and any linter not installed globally).  Can you clarify what you have in mind?  That the Syntastic configuration is included in the project somehow?

**Comment by kevinoid** [comment]
> If your change requires the registry to be fully loaded for certain filetypes this slows down the first run of a checker, sometimes by a few seconds.

Also, to clarify, the `.IsAvailable()` checks are not run.  So the registry isn't "fully" loaded in that sense.  It does require the checkers to be registered so that `.getExec()` can be called to get the executable name.  It also walks the filesystem to check if the executable is present.  This does add overhead which may be unacceptable for some systems/users and could be addressed with an off-by-default configuration.  However, it's nowhere near "a few seconds" for any case I am aware of.

**Comment by lcd047** [comment]
> Can you clarify what you have in mind?
Like I said, I probably don't understand the problem you're trying to solve.  So "clarify" might not be the right word for it.

> That the Syntastic configuration is included in the project somehow?
That's one way to do it: a small Vim script in the base directory of the project, that sets the relevant syntastic variables.  You could source that from an `autocmd`.

Another way to do it would be to write wrapper scripts for the relevant linters.  Those scripts would determine the right executable to run, set the relevant parameters etc. dfepending on the current directory and the path to the file or directory being checked.

There are also plugins that allow you to set variables based on project.  And so on.

**Comment by kevinoid** [comment]
I contribute to enough projects which I don't maintain that using per-project configuration is not workable for me.

I could certainly write a wrapper script to call the appropriate checker executable.  I could even write it in VimL and call it once when the JavaScript filetype is detected and remember the executable path, rather than on each check, and package it as a plugin so I can easily copy it to the places I need it... and end up with exactly what I have.  :smile:

**Comment by jablko** [comment]
[The following](https://github.com/vim-syntastic/syntastic/issues/2099#issuecomment-477734917) worked for me:
```
:let $PATH = system('yarn bin')[:-2] . ':' . $PATH
```

**Comment by lcd047** [comment]
> [The following](https://github.com/vim-syntastic/syntastic/issues/2099#issuecomment-477734917) worked for me:

That's only because you didn't read the thread above.  Which, honestly, you should.  It isn't like none of the people above haven't _thought_ of that, you know.

**Comment by jablko** [comment]
> That's only because you didn't read the thread above. Which, honestly, you should.

Sorry. Do you know if there's a way in Vim to set `$PATH` locally, without updating every other buffer? Analogous to `b:syntastic_shell`?

**Comment by lcd047** [comment]
There is no such thing.  `PATH` is an environment variable, it has nothing to do with buffers.

**Comment by kas** [comment]
I don't want to install ESLint globally. After parsing through this thread, the only solution that worked for me is from this blog post: http://www.spencer.sx/posts/eslint

> let g:syntastic_typescript_checkers=['eslint']
> let g:syntastic_typescript_eslint_exec='node_modules/.bin/eslint'

Note that I was testing opening Vim with the project folder that contains the package.json file. After a while I realized the terminal I ran `vim folder-name` with needed to be at folder-name as well (`cd folder-name`).

**Comment by LvChengbin** [comment]
I wrote a tiny tool for using the local eslint in syntastic, and yarn 2 can also be supported.

https://www.npmjs.com/package/xeslint#syntastic-vim

**Comment by aivils** [comment]
You can easily debug this with setting:

```
:let g:syntastic_debug=3
```
then write the file in the gvim

You can also run test commands
```
:echo syntastic#util#system(' echo $(which npm)')
```
In my case nvm turned out to be the culprit, despite it working fine for years.

It ended up with the following settings:
```
let g:syntastic_javascript_checkers=['eslint']
let g:syntastic_javascript_eslint_generic = 1
let g:syntastic_javascript_eslint_exec = '/bin/ls'
let g:syntastic_javascript_eslint_exe='source "$HOME/.nvm/nvm.sh"; source "$HOME/.nvm/bash_completion"; nvm use; $(npm bin)/eslint'
let g:syntastic_javascript_eslint_args='-f compact'
```
I still don't understand why "gvim" read my ".bashrc" before, but now it doesn't. It happened after the system upgrade.
